### ISSUE TITLE
`spill` is not consistently applied to all traces

### ISSUE THREAD

**Issue by ryzhyk** [feldera/feldera#1785]
I spotted the following fragment in a circuit generated by the `rolling_aggregate` operator: 
![circuit](https://github.com/feldera/feldera/assets/8323135/3753b54f-db84-4aa8-b7c9-dccb553ce7c3)

The two highlighted areas are integrals of the same stream and are redundant. One of them is constructed [here](https://github.com/feldera/feldera/blob/2ffd757be19f1cd9c60c6fc1da420c6ef62ecf2e/crates/dbsp/src/operator/dynamic/time_series/radix_tree/partitioned_tree_aggregate.rs#L268) as part of the `partitioned_tree_aggregate_generic` operator, without calling `spill`. The other one is constructed [here](https://github.com/feldera/feldera/blob/2ffd757be19f1cd9c60c6fc1da420c6ef62ecf2e/crates/dbsp/src/operator/dynamic/time_series/rolling_aggregate.rs#L566). The latter was supposed to return a reference to the same trace, but because of `spill` it ends up building a new trace.

This might also explain why circuits can still use significant amount of memory even with storage enabled.

This isn't hard to fix by adding `spill` in the right place (probably, inside `add_integrate_trace_feedback`), but I am wondering if it would be better to get rid of the `spill` operator instead. It seems too tricky to reason about in combination with caching and I expect we will keep running into issues with it. Ideally, we would encapsulate its functionality inside spine, so that other operators will not have to reason about spilling to disk.

**Comment by gz** [comment]
>but I am wondering if it would be better to get rid of the spill operator instead. It seems too tricky to reason about in combination with caching and I expect we will keep running into issues with it. Ideally, we would encapsulate its functionality inside spine, so that other operators will not have to reason about spilling to disk.

I also think this is likely the better approach since that's what the other LSM systems tend to do rocks/cassandra/scylla etc.

**Comment by blp** [comment]
> > but I am wondering if it would be better to get rid of the spill operator instead. It seems too tricky to reason about in combination with caching and I expect we will keep running into issues with it. Ideally, we would encapsulate its functionality inside spine, so that other operators will not have to reason about spilling to disk.
> 
> I also think this is likely the better approach since that's what the other LSM systems tend to do rocks/cassandra/scylla etc.

Makes sense now that we apply it everywhere.

**Comment by blp** [comment]
> Also we should make the min_storage_rows based on bytes instead of the count

I forked off issue #1812 for that.

**Comment by blp** [comment]
I'm going to try to find out how much of a performance penalty there is if we replace `Vec*` entirely by `Fallback*`. Before, we tried replacing `Vec*` by `File*` and it was a big penalty. But the "fallback" versions should be much cheaper in the case where they don't spill to disk. And, if the performance cost is minimal, then this is much better than just using storage for traces, because it will properly handle big data in individual steps.

**Comment by ryzhyk** [comment]
Does this mean that the `Spillable` traits will no longer be needed (fewer traits is always good news)?

**Comment by blp** [comment]
> Does this mean that the `Spillable` traits will no longer be needed (fewer traits is always good news)?

If it works, then yeah we could get rid of `Spillable` and friends.

**Comment by blp** [comment]
I replaced every `VecIndexedZSet` with `FallbackIndexedZSet`, without enabling storage. For 10M events in Nexmark, the results are:

```
┌───────┬────────────┬───────┬──────────┬─────────┐
│ Query │ #Events    │ Cores │ Fallback │ Vec     │
├───────┼────────────┼───────┼──────────┼─────────┤
│ Q0    │ 10,000,000 │ 16    │ 1.625s   │ 1.668s  │
│ Q1    │ 10,000,000 │ 16    │ 1.435s   │ 1.408s  │
│ Q2    │ 10,000,000 │ 16    │ 1.378s   │ 1.470s  │
│ Q3    │ 10,000,000 │ 16    │ 1.526s   │ 1.534s  │
│ Q4    │ 10,000,000 │ 16    │ 1.741s   │ 1.709s  │
│ Q5    │ 10,000,000 │ 16    │ 1.536s   │ 1.548s  │
│ Q6    │ 10,000,000 │ 16    │ 1.668s   │ 1.671s  │
│ Q7    │ 10,000,000 │ 16    │ 1.826s   │ 1.750s  │
│ Q8    │ 10,000,000 │ 16    │ 1.488s   │ 1.480s  │
│ Q9    │ 10,000,000 │ 16    │ 6.303s   │ 6.260s  │
│ Q12   │ 10,000,000 │ 16    │ 1.468s   │ 1.510s  │
│ Q13   │ 10,000,000 │ 16    │ 1.728s   │ 1.751s  │
│ Q14   │ 10,000,000 │ 16    │ 1.499s   │ 1.421s  │
│ Q15   │ 10,000,000 │ 16    │ 1.618s   │ 1.593s  │
│ Q16   │ 10,000,000 │ 16    │ 4.310s   │ 4.484s  │
│ Q17   │ 10,000,000 │ 16    │ 1.930s   │ 1.976s  │
│ Q18   │ 10,000,000 │ 16    │ 4.278s   │ 4.175s  │
│ Q19   │ 10,000,000 │ 16    │ 4.515s   │ 4.544s  │
│ Q20   │ 10,000,000 │ 16    │ 4.352s   │ 4.341s  │
│ Q21   │ 10,000,000 │ 16    │ 1.490s   │ 1.493s  │
│ Q22   │ 10,000,000 │ 16    │ 1.457s   │ 1.406s  │
└───────┴────────────┴───────┴──────────┴─────────┘
```

In other words, the difference is within the margin of error most of the time.

I'll retry with 100M events. If the pattern holds, I think we can ditch the idea of spilling at all.

**Comment by ryzhyk** [comment]
@blp , what is the value of the spill to file barrier in this experiment?

**Comment by blp** [comment]
For 100M events:

```
┌───────┬─────────────┬───────┬─────────┬──────────┐
│ Query │ #Events     │ Cores │ Vec     │ Fallback │
├───────┼─────────────┼───────┼─────────┼──────────┤
│ Q0    │ 100,000,000 │ 16    │ 14.871s │ 14.606s  │
│ Q1    │ 100,000,000 │ 16    │ 14.056s │ 14.731s  │
│ Q2    │ 100,000,000 │ 16    │ 14.597s │ 14.070s  │
│ Q3    │ 100,000,000 │ 16    │ 14.705s │ 14.930s  │
│ Q4    │ 100,000,000 │ 16    │ 16.857s │ 16.781s  │
│ Q5    │ 100,000,000 │ 16    │ 15.186s │ 15.582s  │
│ Q6    │ 100,000,000 │ 16    │ 16.371s │ 16.776s  │
│ Q7    │ 100,000,000 │ 16    │ 22.490s │ 22.532s  │
│ Q8    │ 100,000,000 │ 16    │ 15.020s │ 14.976s  │
│ Q9    │ 100,000,000 │ 16    │ 91.233s │ 90.584s  │
│ Q12   │ 100,000,000 │ 16    │ 14.552s │ 15.129s  │
│ Q13   │ 100,000,000 │ 16    │ 19.167s │ 20.801s  │
│ Q14   │ 100,000,000 │ 16    │ 14.811s │ 14.447s  │
│ Q15   │ 100,000,000 │ 16    │ 16.976s │ 17.209s  │
│ Q16   │ 100,000,000 │ 16    │ 56.043s │ 54.559s  │
│ Q17   │ 100,000,000 │ 16    │ 21.236s │ 21.444s  │
│ Q18   │ 100,000,000 │ 16    │ 63.783s │ 63.297s  │
│ Q19   │ 100,000,000 │ 16    │ 74.631s │ 73.297s  │
│ Q20   │ 100,000,000 │ 16    │ 68.267s │ 67.449s  │
│ Q21   │ 100,000,000 │ 16    │ 14.922s │ 14.670s  │
│ Q22   │ 100,000,000 │ 16    │ 14.779s │ 13.997s  │
└───────┴─────────────┴───────┴─────────┴──────────┘
```

**Comment by blp** [comment]
> @blp , what is the value of the spill to file barrier in this experiment?

Infinite: storage is disabled. My goal was to figure out the overhead of the extra layer for cases where data doesn't go to storage; that is, the question is, "Is there an unreasonable performance penalty for using Fallback in cases where we don't need to?"

**Comment by ryzhyk** [comment]
My concern is that with storage enabled the relative overhead of spilling transient batches (i.e., batches not in a trace) to disk will be much higher. We should probably measure that overhead before making this change.

**Comment by blp** [comment]
Here is performance for Nexmark q9 with and without making all OrdIndexedZSets into ones that can fall back to storage:

```
                     storage    storage only
min_storage_rows   everywhere    for spine
----------------   ----------   ------------
   1_000_000          6.5 s        6.4 s
     100_000         16.5 s       17.1 s
      50_000         31.5 s       31.2 s
      40_001         40.4 s       40.4 s
      10_000         58.5 s       58.4 s
       1_000        140.8 s      104.9 s
           0        220.6 s      111.7 s
```

The choice of 40001 was because I set `--input-batch-size 40000` on the benchmark.

**Comment by mihaibudiu** [comment]
that's a big range

**Comment by blp** [comment]
> that's a big range

You mean, for min_storage_rows=1000 or less? Yes.

**Comment by blp** [comment]
I am testing a set of changes that get rid of spill and unspill and switch all zsets to ones that can be stored.

**Comment by blp** [comment]
Here are the updated results with the series I'm testing. They're about the same as my [previous test](https://github.com/feldera/feldera/issues/1785#issuecomment-2145777450):

```
                     storage    storage only
min_storage_rows   everywhere    for spine
----------------   ----------   ------------
   1_000_000          6.5 s        6.4 s
     100_000         17.0 s       17.1 s
      50_000         31.1 s       31.2 s
      40_001         38.9 s       40.4 s
      10_000         59.0 s       58.4 s
       1_000        138.6 s      104.9 s
           0        226.6 s      111.7 s
```

**Comment by blp** [comment]
Fixed with PR #1867.
